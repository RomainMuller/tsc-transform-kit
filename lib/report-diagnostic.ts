import { resolve } from 'path';
import { Diagnostic, FormatDiagnosticsHost, formatDiagnosticsWithColorAndContext, sys } from 'typescript';

const FORMAT_HOST: FormatDiagnosticsHost = {
  getCanonicalFileName: resolve,
  getCurrentDirectory: sys.getCurrentDirectory,
  getNewLine: () => sys.newLine,
};

/**
 * Reports multiple `Diagnostic` elements.
 *
 * @param diagnostics the `Diagnostic` objects to be reported.
 * @param writer      the `MessageWriter` function to use for displaying entries. The default writer is `console.error`.
 * @param host        the `FormatDiagnosticHost` to use. The default host uses the standard `typescript.sys` primitives.
 */
export function reportDiagnostics(
  diagnostics: readonly Diagnostic[],
  writer: MessageWriter = console.error,
): void {
  writer(formatDiagnosticsWithColorAndContext(diagnostics, FORMAT_HOST));
}

/**
 * Reports a single `Diagnostic`.
 *
 * @param diagnostic the `Diagnostic` to be reported.
 * @param writer     the `MessageWriter` function to use for displaying.
 * @param host       the `FormatDiagnosticHost` to use.
 */
export function reportDiagnostic(
  diagnostic: Diagnostic,
  writer?: MessageWriter,
): void {
  reportDiagnostics([diagnostic], writer);
}

/**
 * Renders a message to the user.
 *
 * @param message the message to be displayed to the user.
 */
export type MessageWriter = (message: string) => void;
