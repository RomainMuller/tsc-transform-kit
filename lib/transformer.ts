import * as ts from 'typescript';

/**
 * A TypeScript transformer.
 */
export abstract class Transformer<RootNode extends ts.Node> {
  /**
   * The `TransformStage` at which this `Transformer` operates.
   */
  public abstract readonly stages: readonly TransformerStage[];

  /**
   * Performs the TypeScript transformation on the provided `RootNode`.
   *
   * @param program the TypeScript `Program` that is being transformed.
   * @param context the `TransformationContext` under which we are operating.
   * @param node    the `RootNode` to be transformed.
   *
   * @returns the transformed `RootNode`.
   *
   * @internal
   */
  public transform(program: ts.BuilderProgram, context: ts.TransformationContext, node: RootNode): RootNode {
    return ts.visitNode(node, node => this.visit(program, context, node));
  }

  /**
   * Implements the logic for transforming a given TypeScript `node`.
   *
   * @param node the `Node` to be transformed.
   * @param env  the `TransformEnvironment` under which we are operating.
   *
   * @returns a `VisitResult` denoting the outcome of the transformation.
   */
  protected abstract visitNode<T extends ts.Node>(node: T, env: TransformEnvironment): ts.VisitResult<T>;

  private visit<T extends ts.Node>(program: ts.BuilderProgram, context: ts.TransformationContext, node: T): ts.VisitResult<T> {
    let result = this.visitNode(node, { context, program });

    if (result === undefined) {
      return undefined;
    }

    if (!Array.isArray(result)) {
      result = [result];
    }

    result = result.map(newNode => ts.visitEachChild(newNode, toVisit => this.visit(program, context, toVisit), context))
      .filter(result => result != null)
      .map(result => Array.isArray(result) ? result : [result])
      .reduce((acc, newNode) => [...acc, newNode], []);

    return result.length === 1 ? result[0] : result;
  }
}

/**
 * The stages at which transformation can occur.
 */
export const enum TransformerStage {
  /**
   * This transformation happens *before* the *TypeScript* transformations have
   * occurred. Code has not been compiled and it is possible to influence the
   * generated *JavaScript*.
   *
   * In the majority of cases, transformers operate in this phase.
   */
  BEFORE = 'before',

  /**
   * This transformation happens *after* the *TypeScript* transformations have
   * occurred. Code has been compiled already.
   */
  AFTER = 'after',

  /**
   * This transformation happens *after* the *declarations* have been prepared,
   * and type definitions can be altered at this point.
   */
  AFTER_DECLARATIONS = 'afterDeclarations',
}

/**
 * The environment under which a transformation operates.
 */
export interface TransformEnvironment {
  /**
   * The TypeScript `TransformationContext` provided by the compiler framework.
   */
  readonly context: ts.TransformationContext;

  /**
   * The TypeScript `Program`, from which a `TypeChecker` can be obtained.
   */
  readonly program: ts.BuilderProgram;
}
