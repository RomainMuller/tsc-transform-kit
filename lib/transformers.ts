import * as ts from 'typescript';
import {
  Transformer,
  TransformerContext,
  TransformerPhase,
} from './transformer';

export class Transformers {
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  readonly #transformers: Transformer[];

  public constructor(...transformers: readonly Transformer[]) {
    this.#transformers = Array.from(transformers);
  }

  public addTransformer(transformer: Transformer): this {
    this.#transformers.push(transformer);
    return this;
  }

  public forInvalidatedProject<T extends ts.BuilderProgram>(
    project: ts.InvalidatedProject<T>,
  ): ts.CustomTransformers {
    return {
      before: this.#transformers.map((tx) =>
        customTransformerFactory(tx, TransformerPhase.Before),
      ),
      after: this.#transformers.map((tx) =>
        customTransformerFactory(tx, TransformerPhase.After),
      ),
      afterDeclarations: this.#transformers.map((tx) =>
        customTransformerFactory(tx, TransformerPhase.AfterDeclarations),
      ),
    };

    function customTransformerFactory(
      tx: Transformer,
      phase: TransformerPhase,
    ): ts.CustomTransformerFactory {
      return (ctx) => new CustomTransformerFactory(tx, project, ctx, phase);
    }
  }
}

class CustomTransformerFactory<T extends ts.BuilderProgram>
  implements ts.CustomTransformer {
  /* eslint-disable @typescript-eslint/explicit-member-accessibility */
  readonly #delegate: Transformer;
  readonly #transformerContext: TransformerContext;
  /* eslint-enable @typescript-eslint/explicit-member-accessibility */

  public constructor(
    delegate: Transformer,
    invalidatedProject: ts.InvalidatedProject<T>,
    context: ts.TransformationContext,
    phase: TransformerPhase,
  ) {
    this.#delegate = delegate;
    this.#transformerContext = new TransformerContext(
      phase,
      context,
      invalidatedProject,
    );
  }

  public transformBundle(node: ts.Bundle): ts.Bundle {
    return this.#delegate.transform(node, this.#transformerContext);
  }

  public transformSourceFile(node: ts.SourceFile): ts.SourceFile {
    return this.#delegate.transform(node, this.#transformerContext);
  }
}
