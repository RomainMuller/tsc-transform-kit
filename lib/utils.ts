import * as ts from 'typescript';

/**
 * Retrieves the textual name for a `BindingName`.
 *
 * @param name the `BindingName` for which text is needed.
 *
 * @returns the textual expression for the name.
 */
export function getText(name: ts.BindingName | ts.PropertyName): string {
  if (ts.isArrayBindingPattern(name) || ts.isObjectBindingPattern(name)) {
    return getText(name.parent.name);
  } else if (ts.isComputedPropertyName(name)) {
    throw new Error('Cannot find textual representation of ComputedPropertyName elements!');
  }
  return name.text;
}
