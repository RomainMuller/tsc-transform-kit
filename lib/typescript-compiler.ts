import * as ts from 'typescript';
import { Transformer, TransformerStage } from './transformer';
import { reportDiagnostic } from './report-diagnostic';

export class TypeScriptCompiler {
  private readonly transformers: Transformer<ts.Bundle | ts.SourceFile>[] = [];

  public constructor(private readonly configFilePath: string) {}

  public appendTransformer(transformer: Transformer<ts.Bundle | ts.SourceFile>) {
    this.transformers.push(transformer);
  }

  public prependTransformer(transformer: Transformer<ts.Bundle | ts.SourceFile>) {
    this.transformers.unshift(transformer);
  }

  public compileOnce(): ts.EmitResult {
    const parsedCommandLine = this.parseCompilerConfiguration();
    const host = ts.createIncrementalCompilerHost(parsedCommandLine.options);

    const configFileParsingDiagnostics = ts.getConfigFileParsingDiagnostics(parsedCommandLine);
    const program = ts.createIncrementalProgram({
      configFileParsingDiagnostics,
      options: parsedCommandLine.options,
      rootNames: parsedCommandLine.fileNames,
      host,
      projectReferences: parsedCommandLine.projectReferences,
    });

    const transformResult = this.transformProgram(program);
    return {
      ...transformResult,
      diagnostics: [...configFileParsingDiagnostics, ...transformResult.diagnostics],
    };
  }

  public compileAndWatch(watchOptions?: TypeScriptWatchOptions): ts.Watch<ts.BuilderProgram> {
    const host = ts.createWatchCompilerHost(
      this.configFilePath,
      DEFAULT_COMPILER_OPTIONS,
      ts.sys,
      undefined,
      watchOptions?.reportDiagnostic,
      watchOptions?.reportWatchStatus
    );

    const afterProgramCreate = host.afterProgramCreate;
    host.afterProgramCreate = program => {
      const emitResult = this.transformProgram(program);

      for (const diagnostic of emitResult.diagnostics) {
        watchOptions.reportDiagnostic(diagnostic);
      }

      if (watchOptions?.onCompilationComplete != null) {
        watchOptions.onCompilationComplete(emitResult);
      }
      afterProgramCreate?.call(host, program);
    };
    return ts.createWatchProgram(host);
  }

  private parseCompilerConfiguration() {
    const host: ts.ParseConfigFileHost = {
      fileExists: ts.sys.fileExists,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      onUnRecoverableConfigFileDiagnostic: diagnostic => {
        reportDiagnostic(diagnostic);
        throw new Error(`Invalid configuration: ${ts.flattenDiagnosticMessageText(diagnostic.messageText, ts.sys.newLine)}`);
      },
      readDirectory: ts.sys.readDirectory,
      readFile: ts.sys.readFile,
      useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
    };
    const result = ts.getParsedCommandLineOfConfigFile(this.configFilePath, DEFAULT_COMPILER_OPTIONS, host);
    if (result == null) {
      throw new Error(`Unable to parse command line equivalent from the config file at: ${this.configFilePath}`);
    }
    return result;
  }

  private transformProgram(program: ts.BuilderProgram) {
    return program.emit(
      undefined,  // targetSourceFile
      undefined,  // writeFile
      undefined,  // cancellationToken
      false,      // emitOnlyDtsFiles
      {           // customTransformers
        before: this.transformers
          .filter(tx => tx.stage === TransformerStage.BEFORE)
          .map(tx => toTransformerFactory(tx as Transformer<ts.SourceFile>)),
        after: this.transformers
          .filter(tx => tx.stage === TransformerStage.AFTER)
          .map(tx => toTransformerFactory(tx as Transformer<ts.SourceFile>)),
        afterDeclarations: this.transformers
          .filter(tx => tx.stage === TransformerStage.AFTER_DECLARATIONS)
          .map(toTransformerFactory),
      },
    );

    function toTransformerFactory<T extends ts.Bundle | ts.SourceFile>(transformer: Transformer<T>): ts.TransformerFactory<T> {
      return context => node => transformer.transform(program, context, node);
    }
  }
}

export interface TypeScriptCompilerOptions {
  readonly configFilePath: string;
  readonly rootFiles: readonly string[];
}

export interface TypeScriptWatchOptions {
  reportDiagnostic: ts.DiagnosticReporter;
  reportWatchStatus?: ts.WatchStatusReporter;
  onCompilationComplete?: (emitResult: ts.EmitResult) => void;
}

const DEFAULT_COMPILER_OPTIONS: ts.CompilerOptions = {
  alwaysStrict: true,
  charset: 'utf-8',
  composite: true,
  incremental: true,
  inlineSourceMap: true,
  inlineSources: false,
  module: ts.ModuleKind.CommonJS,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  noImplicitAny: true,
  noImplicitReturns: true,
  noImplicitThis: true,
  noUnusedLocals: true,
  noUnusedParameters: true,
  target: ts.ScriptTarget.ES2017,
};
