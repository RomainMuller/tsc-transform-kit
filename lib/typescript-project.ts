import { EventEmitter } from 'events';
import { existsSync } from 'fs';
import * as ts from 'typescript';
import { Transformers } from './transformers';
import { IWatch, Watch } from './watch';

export class TypeScriptProject<T extends ts.BuilderProgram> {
  private readonly system: ts.System;
  private readonly createProgram?: ts.CreateProgram<T>;

  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  readonly #reportDiagnostic: ts.DiagnosticReporter = diag => this.emit(BuildEvent.Diagnostic, diag);
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  readonly #reportSolutionBuilderStatus: ts.DiagnosticReporter = diag => this.emit(BuildEvent.BuildStatus, diag);
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  readonly #reportErrorSummary: ts.ReportEmitErrorSummary = errorCount => this.emit(BuildEvent.ErrorSummary, errorCount);

  private readonly eventEmitter = new EventEmitter();

  /**
   * Creates a new TypeScript project.
   *
   * @param tsconfigPath the path to the tsconfig.json for the project.
   * @param transformers custom transformers to apply.
   * @param options      options for the underlying TypeScript compiler.
   */
  public constructor(
    public readonly tsconfigPath: string,
    public readonly transformers: Transformers = new Transformers(),
    options?: TypeScriptProjectOptions<T>,
  ) {
    if (!existsSync(tsconfigPath)) {
      throw new Error(`${tsconfigPath} does not exist!`);
    }
    this.system = options?.system ?? ts.sys;
    this.createProgram = options?.createProgram;
  }

  /**
   * Builds this TypeScript project once, then returns.
   *
   * @param rootNames         the root names to be compiled.
   * @param defaultOptions    custom build options.
   * @param cancellationToken a cancellation token.
   */
  public build(
    rootNames: string[] = [this.tsconfigPath],
    defaultOptions: ts.BuildOptions = { incremental: true },
    cancellationToken?: ts.CancellationToken,
  ): void {
    const host = ts.createSolutionBuilderHost(
      this.system,
      this.createProgram,
      this.#reportDiagnostic,
      this.#reportSolutionBuilderStatus,
      this.#reportErrorSummary,
    );
    const builder = ts.createSolutionBuilder(host, rootNames, defaultOptions);
    this.consumeBuilder(builder, cancellationToken);
  }

  /**
   * Compiles this project once, then watches for input file changes and
   * re-compiles when needed.
   *
   * @param rootNames         the root names to be compiled.
   * @param defaultOptions    custom build options.
   * @param cancellationToken a cancellation token.
   *
   * @returns an `IWatch` that can be used to stop the process.
   */
  public watch(
    rootNames: string[] = [this.tsconfigPath],
    defaultOptions: ts.BuildOptions = { incremental: true },
    cancellationToken?: ts.CancellationToken,
  ): IWatch {
    const host = ts.createSolutionBuilderHost(
      this.system,
      this.createProgram,
      this.#reportDiagnostic,
      this.#reportSolutionBuilderStatus,
    );
    const createBuilder = () => ts.createSolutionBuilder(host, rootNames, defaultOptions);
    // Work off a new builder, as the tsc API doesn't expose any way to invalidate/revalidate projects
    const watch = new Watch(this.system, watch => this.consumeBuilder(createBuilder(), cancellationToken, watch));

    this.consumeBuilder(createBuilder(), cancellationToken, watch);

    return watch;
  }

  public on(event: BuildEvent.BeforeProject, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public on(event: BuildEvent.AfterProject, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public on(event: BuildEvent.Diagnostic, listener: (diagnostic: ts.Diagnostic) => void): this;
  public on(event: BuildEvent.BuildStatus, listener: (diagnostic: ts.Diagnostic) => void): this;
  public on(event: BuildEvent.ErrorSummary, listener: (errorCount: number) => void): this;
  public on(event: BuildEvent.OutputsGenerated, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public on(event: BuildEvent.OutputsSkipped, listener: (project: ts.InvalidatedProject<T>, reason: OutputsSkippedReason) => void): this;
  // eslint-disable-next-line max-len
  public on(event: BuildEvent.WatchStatus, listener: (diagnostic: ts.Diagnostic, newLine: string, options: ts.CompilerOptions, errorCount: number) => void): this;
  public on(event: BuildEvent, listener: (...any: any[]) => void): this {
    this.eventEmitter.on(event, listener);
    return this;
  }

  public once(event: BuildEvent.BeforeProject, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public once(event: BuildEvent.AfterProject, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public once(event: BuildEvent.Diagnostic, listener: (diagnostic: ts.Diagnostic) => void): this;
  public once(event: BuildEvent.BuildStatus, listener: (diagnostic: ts.Diagnostic) => void): this;
  public once(event: BuildEvent.ErrorSummary, listener: (errorCount: number) => void): this;
  public once(event: BuildEvent.OutputsGenerated, listener: (project: ts.InvalidatedProject<T>) => void): this;
  public once(event: BuildEvent.OutputsSkipped, listener: (project: ts.InvalidatedProject<T>, reason: OutputsSkippedReason) => void): this;
  // eslint-disable-next-line max-len
  public once(event: BuildEvent.WatchStatus, listener: (diagnostic: ts.Diagnostic, newLine: string, options: ts.CompilerOptions, errorCount: number) => void): this;
  public once(event: BuildEvent, listener: (...args: any[]) => void): this {
    this.eventEmitter.once(event, listener);
    return this;
  }

  private consumeBuilder(builder: ts.SolutionBuilder<T>, cancellationToken?: ts.CancellationToken, watch?: Watch): void {
    const next = () => builder.getNextInvalidatedProject(cancellationToken);
    for (let invalidatedProject = next(); invalidatedProject != null; invalidatedProject = next()) {
      this.emit(BuildEvent.BeforeProject, invalidatedProject);
      try {
        const exitStatus = invalidatedProject.done(
          cancellationToken,
          this.system.writeFile,
          this.transformers.forInvalidatedProject(invalidatedProject),
        );
        switch (exitStatus) {
          case ts.ExitStatus.Success:
            this.emit(BuildEvent.OutputsGenerated, invalidatedProject);
            break;
          case ts.ExitStatus.DiagnosticsPresent_OutputsGenerated:
            this.emit(BuildEvent.OutputsGenerated, invalidatedProject);
            break;
          case ts.ExitStatus.DiagnosticsPresent_OutputsSkipped:
            this.emit(BuildEvent.OutputsSkipped, invalidatedProject, OutputsSkippedReason.DiagnosticsPresent);
            break;
          case ts.ExitStatus.InvalidProject_OutputsSkipped:
            this.emit(BuildEvent.OutputsSkipped, invalidatedProject, OutputsSkippedReason.InvalidProject);
            break;
          case ts.ExitStatus.ProjectReferenceCycle_OutputsSkipped:
          case ts.ExitStatus.ProjectReferenceCycle_OutputsSkupped:
            this.emit(BuildEvent.OutputsSkipped, invalidatedProject, OutputsSkippedReason.ProjectReferenceCycle);
            break;
          default:
            throw new Error(`Unsupported exitStatus: ${ts.ExitStatus[exitStatus]}`);
        }
      } finally {
        if (watch != null) {
          watch.watchConfigFile(invalidatedProject.project);
        }
        this.emit(BuildEvent.AfterProject, invalidatedProject);
      }
    }
  }

  private emit(event: BuildEvent.BeforeProject, project: ts.InvalidatedProject<T>): boolean;
  private emit(event: BuildEvent.AfterProject, project: ts.InvalidatedProject<T>): boolean;
  private emit(event: BuildEvent.Diagnostic, diagnostic: ts.Diagnostic): boolean;
  private emit(event: BuildEvent.BuildStatus, diagnostic: ts.Diagnostic): boolean;
  private emit(event: BuildEvent.ErrorSummary, errorCount: number): boolean;
  private emit(event: BuildEvent.OutputsGenerated, project: ts.InvalidatedProject<T>): boolean;
  private emit(event: BuildEvent.OutputsSkipped, project: ts.InvalidatedProject<T>, reason: OutputsSkippedReason): boolean;
  private emit(event: BuildEvent.WatchStatus, diagnostic: ts.Diagnostic, newLine: string, options: ts.CompilerOptions, errorCount: number): boolean;
  private emit(event: BuildEvent, ...args: any): boolean {
    return this.eventEmitter.emit(event, ...args);
  }
}

/**
 * Custom options to initialize a TypeScript compiler.
 */
export interface TypeScriptProjectOptions<T extends ts.BuilderProgram = ts.EmitAndSemanticDiagnosticsBuilderProgram> {
  /**
   * A TypeScrypt system.
   */
  readonly system?: ts.System;

  /**
   * A TypeScript createProgram function.
   */
  readonly createProgram?: ts.CreateProgram<T>;
}

/**
 * Events that a TypeScript project emits during build.
 */
export const enum BuildEvent {
  /** Happens before a project is transformed. */
  BeforeProject = 'before',
  /** Happens after a project is transformed. */
  AfterProject = 'after',

  /** Happens whenever a diagnostic message is generated. */
  Diagnostic = 'diagnostic',
  /** Happens when the status of a build changes. */
  BuildStatus = 'buildStatus',
  /** Happens when an error summary is generated (after a build). */
  ErrorSummary = 'errorSummary',

  /** Emitted when a project was transformed and generated outputs. */
  OutputsGenerated = 'outputsGenerated',
  /** Happens when a project was transformed, but outputs were not generated. */
  OutputsSkipped = 'outputsSkipped',

  /** Happens whent he status of a watch changes. */
  WatchStatus = 'watchStatus',
}

/**
 * Reasons for why outputs were not emitted.
 */
export const enum OutputsSkippedReason {
  /** A diagnositic message (usually an error) prevented generation. */
  DiagnosticsPresent = 'diagnosticsPresent',
  /** The project's configuration is invalid. */
  InvalidProject = 'invalidProject',
  /** A reference cycle involves the project. */
  ProjectReferenceCycle = 'projectReferenceCycle',
}
