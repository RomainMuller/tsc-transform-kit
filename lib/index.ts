import * as utils from './utils';

export * from './transformer';
export * from './typescript-compiler';
export { utils };
