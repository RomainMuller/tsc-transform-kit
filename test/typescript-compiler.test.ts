import * as fs from 'fs-extra';
import { join } from 'path';
import * as ts from 'typescript';
import { TypeScriptCompiler } from '../lib';
import { TsConfigJson, withProject, withProjectSync } from './with-project';

const tsconfig: TsConfigJson = {
  compilerOptions: {
    alwaysStrict: true,
    charset: 'utf-8',
    composite: true,
    declaration: true,
    incremental: true,
    inlineSourceMap: true,
    module: ts.ModuleKind.CommonJS,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    noImplicitAny: true,
    noImplicitReturns: true,
    noImplicitThis: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    strict: true,
    target: ts.ScriptTarget.ES2018,
  }
};

describe('compileOnce', () => {
  test('successfully compiles project', () => {
    withProjectSync(tsconfig, projectRoot => {
      fs.writeFileSync(join(projectRoot, 'index.ts'), INDEX_TS, { encoding: 'utf-8' });

      const compiler = new TypeScriptCompiler(join(projectRoot, 'tsconfig.json'));

      const result = compiler.compileOnce();
      expect(result.emitSkipped).toBeFalsy();
      expect(result.diagnostics).toEqual([]);

      expect(fs.readFileSync(join(projectRoot, 'index.js'), { encoding: 'utf-8' }))
        .toMatchSnapshot('compiler JS output');
      expect(fs.readFileSync(join(projectRoot, 'index.d.ts'), { encoding: 'utf-8' }))
        .toMatchSnapshot('compiler DTS output');
    });
  });
});

describe('compileAndWatch', () => {
  test('successfully compiles project', () =>
    withProject(tsconfig, async projectRoot => {
      fs.writeFileSync(join(projectRoot, 'index.ts'), INDEX_TS, { encoding: 'utf-8' });

      const compiler = new TypeScriptCompiler(join(projectRoot, 'tsconfig.json'));

      let ok!: () => void;
      let ko!: (reason: any) => void;
      const checksPassed = new Promise<void>((resolve, reject) => { ok = resolve; ko = reject; });

      let cycle = 0;
      const watch = compiler.compileAndWatch({
        reportWatchStatus: () => /* ignore */ null,
        reportDiagnostic: () => /* ignore */ null,
        onCompilationComplete: result => {
          try {
            expect(result.emitSkipped).toBeFalsy();
            expect(result.diagnostics).toEqual([]);

            const js = fs.readFileSync(join(projectRoot, 'index.js'), { encoding: 'utf-8' });
            const dts = fs.readFileSync(join(projectRoot, 'index.d.ts'), { encoding: 'utf-8' });

            if (cycle === 0) {
              expect(js).toContain('TestClass');
              expect(dts).toContain('TestClass');

              fs.writeFileSync(join(projectRoot, 'index.ts'), 'export enum TestEnum { ONE, TWO }', { encoding: 'utf-8' });

              cycle++;
            } else {
              expect(js).toContain('TestEnum');
              expect(dts).toContain('TestEnum');

              ok();
            }
          } catch (err) {
            ko(err);
          }
        },
      });

      try {
        await checksPassed;
      } finally {
        watch.close();
      }
    })
  );
});

const INDEX_TS = `
export class TestClass {
  public readonly property = 1337;
}
`;
