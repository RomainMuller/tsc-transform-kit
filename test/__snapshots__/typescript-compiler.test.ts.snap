// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileOnce successfully compiles project: compiler DTS output 1`] = `
"export declare class TestClass {
    readonly property = 1337;
}
"
`;

exports[`compileOnce successfully compiles project: compiler JS output 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
class TestClass {
    constructor() {
        this.property = 1337;
    }
}
exports.TestClass = TestClass;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJpbmRleC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUNBLE1BQWEsU0FBUztJQUF0QjtRQUNrQixhQUFRLEdBQUcsSUFBSSxDQUFDO0lBQ2xDLENBQUM7Q0FBQTtBQUZELDhCQUVDIn0="
`;
